controllers/rollback_handlers.go

```diff
@@ -18,18 +18,90 @@ package controllers
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
 //nolint:unparam
-func (r *ImageBasedUpgradeReconciler) handleRollback(ctx context.Context, ibu *lcav1alpha1.ImageBasedUpgrade) (ctrl.Result, error) {
+func (r *ImageBasedUpgradeReconciler) startRollback(ctx context.Context, ibu *lcav1alpha1.ImageBasedUpgrade) (ctrl.Result, error) {
+	utils.SetRollbackStatusInProgress(ibu, "Initiating rollback")
+
+	stateroot, err := r.RPMOstreeClient.GetUnbootedStaterootName()
+	if err != nil {
+		utils.SetRollbackStatusFailed(ibu, err.Error())
+		return requeueWithError(err)
+	}
+
+	if _, err := r.Executor.Execute("mount", "/sysroot", "-o", "remount,rw"); err != nil {
+		return requeueWithError(err)
+	}
+
+	stateRootRepo := common.PathOutsideChroot(ostreeclient.StaterootPath(stateroot))
+
+	// Save the CR for post-reboot restore, stripping Upgrade conditions and ResourceVersion from IBU CR before saving
+	r.Log.Info("Save the IBU CR to the old state root before pivot")
+	filePath := filepath.Join(stateRootRepo, utils.IBUFilePath)
+	// Temporarily empty resource version so the file can be used to restore status
+	rv := ibu.ResourceVersion
+	ibu.ResourceVersion = ""
+	if err := lcautils.MarshalToFile(ibu, filePath); err != nil {
+		return requeueWithError(err)
+	}
+	ibu.ResourceVersion = rv
+
+	r.Log.Info("Finding unbooted deployment")
+	deploymentIndex, err := r.RPMOstreeClient.GetUnbootedDeploymentIndex()
+	if err != nil {
+		utils.SetRollbackStatusFailed(ibu, err.Error())
+		return requeueWithError(fmt.Errorf("failed to get unbooted deployment index: %w", err))
+	}
+
+	// Set the new default deployment
+	r.Log.Info("Checking for set-default feature")
+
+	if r.OstreeClient.IsOstreeAdminSetDefaultFeatureEnabled() {
+		r.Log.Info("set-default feature available")
 
-	// TODO actual steps
-	// If completed, update conditions and return doNotRequeue
+		if err = r.OstreeClient.SetDefaultDeployment(deploymentIndex); err != nil {
+			utils.SetRollbackStatusFailed(ibu, err.Error())
+			return requeueWithError(fmt.Errorf("failed to set default deployment for pivot: %w", err))
+		}
+	} else {
+		r.Log.Info("set-default feature not available")
+
+		// Check to make sure the default deployment is set
+		if deploymentIndex != 0 {
+			msg := "default deployment must be manually set for next boot"
+			utils.SetRollbackStatusFailed(ibu, msg)
+			return requeueWithError(fmt.Errorf(msg))
+		}
+	}
+
+	// Write an event to indicate reboot attempt
+	r.Recorder.Event(ibu, corev1.EventTypeNormal, "Reboot", "System will now reboot")
+	err = r.rebootToNewStateRoot()
+	if err != nil {
+		//todo: abort handler? e.g delete desired stateroot
+		r.Log.Error(err, "")
+		utils.SetUpgradeStatusFailed(ibu, err.Error())
+		return doNotRequeue(), nil
+	}
+
+	return doNotRequeue(), nil
+}
+
+//nolint:unparam
+func (r *ImageBasedUpgradeReconciler) finishRollback(ctx context.Context, ibu *lcav1alpha1.ImageBasedUpgrade) (ctrl.Result, error) {
 	utils.SetStatusCondition(&ibu.Status.Conditions,
```

Comment:
nit: You have SetRollbackStatusCompleted() defined, but looks like you are not using it.